elastic/kibana
# [Response Ops][Alerting] Update stack rules to respect max alert limit

Resolves https://github.com/elastic/kibana/issues/138311

## Summary

Updates index threshold and ES query rule types to use the alerting circuit breaker services.

### ES Query

The ES query rule type only ever generates one alert: `query matched` so no changes were needed to the executor. I just updated it to get the limit and report back that the limit is not reached.

### Index Threshold

The index threshold rule type created a group by aggregation to get the count or metric value of a configured field. Once it gets all those values back, it would compare each value against the configured threshold to determine whether an alert should be created. I updated the query to use a `bucket_selector` aggregation to apply the configured threshold within the ES request so the only data returned from the request should be results that already meet the condition. This allows us to limit the number of results returned to ensure it is less than the alert limit.

This does have the side effect of eliminating information about the value of recovered alerts. Previously, if a group has a value that does not meet the condition, we were getting that information anyway in our ES results and adding it to the recovered context (for example, `host-1` has avg cpu usage of 0.6 which is less than the threshold of 0.9). Now, we are no longer getting that information, so the context will show `unknown` (for example, we know that `host-1` has recovered and its avg cpu usage is less than the threshold of 0.9 but we don't know the exact value).

Note: We have an existing FT that tests the circuit breaker for index threshold rules in `x-pack/test/alerting_api_integration/spaces_only/tests/alerting/builtin_alert_types/circuit_breaker/index_threshold_max_alerts.ts`. The difference between using the limit services and not using the limit services is that the rule type can finish any post-processing after alert creation and return rule state to be persisted. For this particular rule type, the only thing that happens after alert creation is recovery alert generation and no rule state is returned. This is why the existing FT still works even after these changes.

## To Verify

1. Set the alert limit lower in the `kibana.yml` file (I used 20)
2. I used [es-apm-sys-sim](https://github.com/pmuellr/es-apm-sys-sim) to start generating data. I started with 5 hosts
Create an index threshold rule that groups by host name and always alerts (CPU pct > 0). Attach a server log action that notifies on every active alert.
3. See rule run and generate 5 alerts.
4. Re-run es-apm-sys-sim with 30 hosts.
5. The next run of the rule should hit the circuit breaker. You should see the warning in the UI and the number of alerts should max out at 20

There are some useful console logs inside `x-pack/plugins/stack_alerts/server/alert_types/index_threshold/rule_type.ts` and `x-pack/plugins/triggers_actions_ui/server/data/lib/time_series_query.ts` that you can uncomment so that you can see the request going to ES and the responses returned. You can compare the responses between this branch and running in `main` to see that the data returned from ES has fewer buckets.

### Checklist
